File: src/main/java/net/xalcon/technomage/Technomage.java
package net.xalcon.technomage;

import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.SidedProxy;
import net.minecraftforge.fml.common.event.*;
import net.minecraftforge.fml.common.network.NetworkRegistry;
import net.xalcon.technomage.common.GuiHandlerTechnomage;
import net.xalcon.technomage.common.IProxy;
import net.xalcon.technomage.common.commands.CommandTechnomage;
import net.xalcon.technomage.common.init.TMDimensions;
import net.xalcon.technomage.common.init.TMRecipes;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

@Mod
(
        modid = Technomage.MOD_ID,
        name = Technomage.NAME,
        version = Technomage.VERSION,
        acceptedMinecraftVersions = Technomage.MC_VERSION,
        dependencies = Technomage.DEPENDENCIES
)
public class Technomage
{
    public final static String MOD_ID = "technomage";
    public final static String NAME = "Technomage";
    public final static String GROUP = "net.xalcon." + MOD_ID;
    public final static String VERSION = "@VERSION@";
    public final static String MC_VERSION = "1.12";
    public final static String DEPENDENCIES = "required-after:forge@[14.21.0.2328,)";

    private static Technomage instance = new Technomage();

    @SidedProxy(clientSide = GROUP + ".client.ClientProxy", serverSide = GROUP + ".server.ServerProxy")
    public static IProxy Proxy;

    public final static Logger Log = LogManager.getLogger(MOD_ID);

    @Mod.EventHandler
    public static void onConstruction(FMLConstructionEvent event)
    {
    }

    @Mod.EventHandler
    public static void onServerStarting(FMLServerStartingEvent event)
    {
        event.registerServerCommand(new CommandTechnomage());
    }

    @Mod.EventHandler
    public static void onPreInit(FMLPreInitializationEvent event)
    {
        TMDimensions.init();
        Proxy.preInit(event);
        NetworkRegistry.INSTANCE.registerGuiHandler(instance, new GuiHandlerTechnomage());
    }

    @Mod.EventHandler
    public static void onInit(FMLInitializationEvent event)
    {
        Proxy.init(event);
        TMRecipes.init();
    }

    @Mod.EventHandler
    public static void onPostInit(FMLPostInitializationEvent event)
    {
        Proxy.postInit(event);
    }

    @Mod.InstanceFactory
    public static Technomage getInstance()
    {
        return instance;
    }
}

File: src/main/kotlin/net/xalcon/technomage/Technomage.kt
package net.xalcon.technomage

import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.SidedProxy
import net.minecraftforge.fml.common.event.FMLConstructionEvent
import net.minecraftforge.fml.common.event.FMLInitializationEvent
import net.minecraftforge.fml.common.event.FMLPostInitializationEvent
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent
import net.minecraftforge.fml.common.network.NetworkRegistry
import net.xalcon.technomage.common.CommonProxy
import net.xalcon.technomage.common.TechnomageGuiHandler
import org.apache.logging.log4j.LogManager
import kotlin.experimental.and

@Mod
(
        modid = Technomage.MOD_ID,
        name = Technomage.NAME,
        version = Technomage.VERSION,
        acceptedMinecraftVersions = Technomage.MC_VERSION,
        dependencies = Technomage.DEPENDENCIES
)
object Technomage
{
    const val MOD_ID = "technomage"
    const val NAME = "Technomage"
    const val GROUP = "net.xalcon." + MOD_ID
    const val VERSION = "@VERSION@"
    const val MC_VERSION = "1.11.2"
    const val DEPENDENCIES = "required-after:forge@[13.20.0.2304,);required-after:forgelin@[1.4.0,)"

    @SidedProxy(clientSide = GROUP + ".client.ClientProxy", serverSide = GROUP + ".server.ServerProxy")
    lateinit var Proxy : CommonProxy

    val Log = LogManager.getLogger(MOD_ID)

    val x:Byte = -1
    var y:Int = 0

    @Mod.EventHandler
    fun onConstruction(event:FMLConstructionEvent)
    {
        y = (x and 0xFF.toByte()).toInt()
    }

    @Mod.EventHandler
    fun onPreInit(event:FMLPreInitializationEvent)
    {
        Proxy.preInit(event)
        TMBlocks.init()
        TMItems.init()

        NetworkRegistry.INSTANCE.registerGuiHandler(this, TechnomageGuiHandler)
    }

    @Mod.EventHandler
    fun onInit(event:FMLInitializationEvent)
    {
    }

    @Mod.EventHandler
    fun onPostInit(event:FMLPostInitializationEvent)
    {
    }

    @JvmStatic
    @Mod.InstanceFactory
    fun createInstance() = Technomage
}